extensions/vi-mode/options.lisp

```diff
@@ -0,0 +1,178 @@
+                  (handler-case (parse-number new-value)
+                    (error () new-value)))
+                 (string
+                  new-value)
+                 (otherwise new-value))))
+        (t
+         (assert (and (null prefix)
+                      (null suffix)))
+         (setf (vi-option-value option) t))))))
+
+(defmacro define-vi-option (name (default &key (type t) aliases) &rest others)
+  (once-only (default)
+    (with-gensyms (option alias new-value)
+      (let ((name-str (string-downcase name)))
+        `(progn
+           (check-type ,default ,type)
+           (dolist (,alias ',aliases)
+             (setf (gethash ,alias *option-aliases*) ,name-str))
+           (let ((,option
+                   (make-vi-option :name ,name-str
+                                   :%value ,default
+                                   :default ,default
+                                   :type ',type
+                                   :aliases ',aliases
+                                   :getter ,(when-let (getter-arg (find :getter others :key #'car))
+                                              `(lambda () ,@(rest getter-arg)))
+                                   :set-hook ,(when-let (set-hook-arg (find :set-hook others :key #'car))
+                                                `(lambda ,@(rest set-hook-arg)))
+                                   :documentation ,(when-let (doc-arg (find :documentation others :key #'car))
+                                                     (second doc-arg)))))
+             (setf (gethash ,name-str *options*) ,option)
+             (defun ,name ()
+               (vi-option-value (get-option ,name-str)))
+             (defun (setf ,name) (,new-value)
+               (setf (vi-option-value (get-option ,name-str)) ,new-value))))))))
+
+(defun auto-change-directory (buffer-or-window)
+  (change-directory (etypecase buffer-or-window
+                      (lem:buffer (lem:buffer-directory buffer-or-window))
+                      (lem:window (lem:buffer-directory (lem:window-buffer buffer-or-window))))))
+
+(define-vi-option autochdir (nil :type boolean :aliases ("acd"))
+  (:documentation "A flag on whether change the current directory to the buffer directory automatically.
+  Default: nil
+  Aliases: acd")
+  (:set-hook (new-value)
+   (if new-value
+       (progn
+         (lem:add-hook lem:*find-file-hook* 'auto-change-directory)
+         (dolist (window (lem:window-list))
+           (lem:add-hook (lem-core::window-switch-to-buffer-hook window) 'auto-change-directory)
+           (lem:add-hook (lem-core:window-leave-hook window) 'auto-change-directory)))
+       (progn
+         (lem:remove-hook lem:*find-file-hook* 'auto-change-directory)
+         (dolist (window (lem:window-list))
+           (lem:remove-hook (lem-core::window-switch-to-buffer-hook window) 'auto-change-directory)
+           (lem:remove-hook (lem-core:window-leave-hook window) 'auto-change-directory))))))
```

Comment:
Perhaps there is an issue that does not apply to windows that are split after this.
It might be a good idea to have the issue.

Reply:
I tried some cases on my machine, but it seems to be working. I don't know why.